### PR TITLE
[CallableWrapper] Add CallableWrapper component and framework integration

### LINKED ISSUE #57079: [HttpKernel] Add before and after hooks to controller actions
### Description

When we need to execute code before or after controllers, we need to:

1. create a listener class and register the event listeners
2. in the event listener, check if the incoming request is the master request
3. check manually the route name or if the controller::action matches the expected names
4. execute the business code 

Creating such listener once is ease and works properly. But, the down side of this approach is that we don't have much clarity on what is being executed before or after controllers, mainly business code. (Framework code is ok).

The suggestion is to add two more attributes to the framework where we can declare what hooks will be executed before and/or after controller actions.

WDYT?

### Example

The following example will trigger `Stopwatch::__invoke` service before/after all actions in this controller.

```php

<?php

#[Before(Stopwatch::class, ["dev"])]
#[After(Stopwatch::class, ["dev"])]
class MyController {
    public function fooAction(): Response {
        return new Response('foo');
    }

    public function barAction(): Response {
        return new Response('foo');
    }
}
```
Or in case to add hooks to a specific controller action:

```php

<?php

class MyController {

    #[Before(service: 'my_stop_watch_service', when: ["dev"])]
    #[After(service: 'my_stop_watch_service', when: ["dev"])]
    public function fooAction(): Response {
        return new Response('foo');
    }
}
```

This new feature suggestion is to controller actions but could be expanded to console commands as well.

This is not the case to change the request/command workflow. All of this can be done by the native event dispatcher.

### PR DESCRIPTION
**Pull request by yceruto**
| Q             | A
| ------------- | ---
| Branch?       | 7.3
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Issues        | Fix https://github.com/symfony/symfony/issues/57079
| License       | MIT

This is an attempt to solve the linked issue in a generic way, making it applicable in other areas where it might be useful as well.

> [!NOTE]
> Inspired by [Python decorators](https://peps.python.org/pep-0318/)

This component implements the [Decorator Pattern](https://en.wikipedia.org/wiki/Decorator_pattern) around any [PHP callable](https://www.php.net/manual/en/language.types.callable.php), allowing you to:
* Execute logic before or after a callable is executed
* Skip the execution of a callable by returning earlier
* Modify the result of a callable

It solves issues related to subclass explosion and inflexibility in inheritance-based designs. The pattern is useful for extending object behavior without modifying the original code, making it ideal for scenarios where different combinations of behaviors are needed.

The component contains two main building blocks:
 1. The `CallableWrapperInterface` adapter (the wrapper): contains the wrapping implementation, essentially defining what should be done before and after the targeted callable is executed.
 2. The wrapper attribute which must extend from `CallableWrapperAttribute` or `CallableWrapperAttributeInterface`: links a callable to a wrapper and collects its options if needed.

Example:
```php
#[\Attribute(\Attribute::TARGET_METHOD)]
class Debug extends CallableWrapperAttribute
{
}

class DebugCallableWrapper implements CallableWrapperInterface
{
    public function wrap(\Closure $func): \Closure
    {
        return function (mixed ...$args) use ($func): mixed {
            echo "Do something before\n";

            $result = $func(...$args);

            echo "Do something after\n";

            return $result;
        };
    }
}
```
```php
#[AsController]
class GreetingController
{
    #[Debug]
    public function hello(string $name): void
    {
        echo "Hello $name\n"
    }
}
```
```php
#[AsMessageHandler]
class GreetingMessageHandler
{
    #[Debug]
    public function __invoke(Message $message): void
    {
        echo "Hello $message->name \n"
    }
}
```
The `Debug` attribute holds all metadata needed for the linked wrapper, which must be referenced using the `CallableWrapperAttributeInterface::wrappedBy()` method. By default, the abstract `CallableWrapperAttribute` class returns `static::class.'CallableWrapper'` as convention.

Wrappers can be nested, so the order in which you define them around the targeted callable really matters. This is a visual representation of multiple wrappers around a function and how they wrap around each other:

image

In short, the closer a wrapper is to the targeted callable, the higher its execution priority.

Callable wrappers might require some options. To handle this, you can define them using the constructor method and add the metadata attribute as a new argument in your wrapper implementation:

```php
#[\Attribute(\Attribute::TARGET_METHOD)]
class Debug extends CallableWrapperAttribute
{
    public function __construct(
        public readonly string $prefix = '',
    ) {
    }
}
```
```php
class DebugCallableWrapper implements CallableWrapperInterface
{
    public function wrap(\Closure $func, Debug $debug = new Debug()): \Closure
    {
        // do something with $debug->prefix
    }
}
```

Final usage:
```php
class GreetingHandler
{
    #[Debug(prefix: 'say: ')]
    public function hello(string $name): void
    {
        echo "Hello $name\n"
    }
}
```
The component requires a middleware layer to wrap the targeted callable before it's actually called. So, after the framework integration, a service named `callable_wrapper` -> `CallableWrapper` -> aliasing to `CallableWrapperInterface` will be available with all collected wrapper services tagged with `callable_wrapper` (see `CallableWrapperPass`). Then, you'll do:
```php
# In any middleware layer
$wrapped = $callableWrapper->wrap($callable(...)); // to apply the wrappers around that callable
$wrapped(...$args); // to execute it
```

| Pros | Cons |
| --- | --- |
| ✅ You can extend an object's behavior without making a new subclass | ❌ It’s hard to remove a specific wrapper from the wrappers stack |
| ✅ You can combine several behaviors by wrapping a `callable` into multiple wrappers | ❌ It’s hard to implement a wrapper in such a way that its behavior doesn’t depend on the order in the wrapper stack |
| ✅ Single Responsibility Principle. You can divide a monolithic `callable` that implements many possible variants of behavior into several smaller `callable`.

Refer to the tests for more details on use cases and how wrappers are executed in the final stage.

> [!IMPORTANT]
> Why create a new component? Because it can be implemented across various components of our Symfony project and application, such as HttpKernel/Controllers, Messenger/Handlers, and potentially other custom producer/consumer mechanisms that contain a middleware layer.

Cheers!

### CODE REVIEW

File: src/Symfony/Component/Decorator/CHANGELOG.md [hunk 1]
```diff
@@ -0,0 +1,7 @@
+CHANGELOG
+=========
+
+7.2
+---
+
+ * Add the component
```
Comment (OskarStark) [inline src/Symfony/Component/Decorator/CHANGELOG.md]:
```suggestion
 * Add the component as experimental
```

**Comment by 94noni** [comment]
Hello @yceruto 
Interesting feature, may i ask you if such component can be used to leverage thing like this  
https://github.com/symfony/symfony/issues/47184
thank you

File: src/Symfony/Component/Decorator/README.md [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+{
+    #[Debug]
+    public function hello(string $name): void
+    {
+        echo "Hello $name\n"
+    }
+}
+
+$greeting = new Greeting();
+$decorator->call($greeting->hello(...), 'Fabien');
```
Comment (94noni) [inline src/Symfony/Component/Decorator/README.md]: Perhaps write the output to give full example?

File: src/Symfony/Component/Decorator/README.md [hunk 3]
```diff
@@ -0,0 +1,72 @@
+Clock Component
```
Comment (94noni) [inline src/Symfony/Component/Decorator/README.md]: Decorator :)

File: src/Symfony/Component/Decorator/composer.json [hunk 4]
```diff
@@ -0,0 +1,34 @@
+{
+    "name": "symfony/decorator",
+    "type": "library",
+    "description": "Decorate the functionality of other functions",
```
Comment (94noni) [inline src/Symfony/Component/Decorator/composer.json]:
```suggestion
    "description": "Decorate the functionality of other codes",
```

?

Comment (yceruto) [reply src/Symfony/Component/Decorator/composer.json]: `codes` is too broad, `functions` is a bit more specific and aligned to the concept.

File: src/Symfony/Component/Decorator/README.md (hunk 3 above)
Comment (yceruto) [reply src/Symfony/Component/Decorator/README.md]: Oops! :)

File: src/Symfony/Component/Decorator/README.md (hunk 2 above)
Comment (yceruto) [reply src/Symfony/Component/Decorator/README.md]: Output added, thanks!

**Comment by yceruto** [comment]
@94noni In principle, yes, but not by default for every service. It would require a middleware layer to hook into it before the method is called. Currently, only proxies can do that for services. Somehow, collect the class/method decorators for each service and call them from within? but not sure about performance though.

**Comment by yceruto** [comment]
✅ Added Framework integration and controller decoration support

✅ As well as a new Doctrine transaction decorator through `#[Transactional]` attribute. This decorator can easily help decouple the business logic layer from the infrastructure/persistence layer in any layered-arch app. For example, using the collection-oriented Repository pattern, we can write an API endpoint like this one:

```php
#[Route('/tasks', methods: 'POST')]
class CreateTaskController
{
    public function __construct(private TaskRepositoryInterface $repository)
    {
    }

    #[Serialize(format: 'json')]
    #[Transactional]
    public function __invoke(#[MapRequestPayload] TaskPayload $payload): Task
    {
        $task = new Task($payload->description, $payload->dueDate);

        $this->repository->add($task); // $this->entityManager->persist($task);

        return $task;
    }
}
```
Keeping the controller code free of persistence and presentation deps. The `#[Serialize]` decorator can be used as a `Serializer` (food for another PR), which serializes the result of the given function according to the request format.

This is just an example. There are more approaches and architectures where decorators can be useful.

File: src/Symfony/Component/Decorator/DecoratorInterface.php [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+/**
+ * Decorates the functionality of other function.
+ *
+ * @author Yonel Ceruto <open@yceruto.dev>
+ *
+ * @experimental
+ */
+interface DecoratorInterface
+{
+    public function decorate(\Closure $func): \Closure;
```
Comment (alexandre-daubois) [inline src/Symfony/Component/Decorator/DecoratorInterface.php]:
```suggestion
    public function decorate(\Closure $decorated): \Closure;
```

Maybe?

File: src/Symfony/Component/Decorator/Attribute/Decorate.php [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+namespace Symfony\Component\Decorator\Attribute;
+
+/**
+ * An attribute that specifies under which decorator a function or method should be wrapped.
+ *
+ * @author Yonel Ceruto <open@yceruto.dev>
+ *
+ * @experimental
+ */
+#[\Attribute(\Attribute::TARGET_FUNCTION | \Attribute::TARGET_METHOD | \Attribute::IS_REPEATABLE)]
```
Comment (alexandre-daubois) [inline src/Symfony/Component/Decorator/Attribute/Decorate.php]: What about accepting classes also? I see two possibilities:
- Only support callable classes (with `__invoke()` their **only** public method)
- Plug on every method.

We could imagine a repository decorated with some stopwatches to spy on execution times of each query builders. Having the possibility to define the decorator on the class would be nice instead of repeating everything (and we know Doctrine repo are not the smallest classes...).

**Review by alexandre-daubois — COMMENTED** [review]
I'm not a big fan on putting `readonly` on each and every class. I'm not sure it does make much sense on services, it makes more sense on DTO/VOs to me. But that's just my two cents. :slightly_smiling_face: 

However, I like very much the idea behind this component. Thank you for proposing it!

File: src/Symfony/Component/Decorator/DecoratorInterface.php (hunk 5 above)
Comment (yceruto) [reply src/Symfony/Component/Decorator/DecoratorInterface.php]: At this point it's not decorated yet :)

File: src/Symfony/Component/Decorator/Attribute/Decorate.php (hunk 6 above)
Comment (yceruto) [reply src/Symfony/Component/Decorator/Attribute/Decorate.php]: Good idea! I'm keeping it simple for this proposal, but there are a lot of feature around it. Food for future PRs 👍

File: src/Symfony/Bridge/Doctrine/Decorator/DoctrineTransactionDecorator.php [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+            try {
+                $return = $func(...$args);
+
+                $entityManager->flush();
+                $entityManager->getConnection()->commit();
+
+                return $return;
+            } catch (\Throwable $e) {
+                $entityManager->close();
+                $entityManager->getConnection()->rollBack();
```
Comment (alexandre-daubois) [inline src/Symfony/Bridge/Doctrine/Decorator/DoctrineTransactionDecorator.php]: Shouldn't the connection rollback happen before the entity manager is closed?

File: src/Symfony/Bridge/Doctrine/Tests/Decorator/DoctrineTransactionDecoratorTest.php [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+    {
+        $this->connection->expects($this->once())
+            ->method('beginTransaction')
+        ;
+        $this->connection->expects($this->once())
+            ->method('commit')
+        ;
+        $this->entityManager->expects($this->once())
+            ->method('flush')
+        ;
```
Comment (alexandre-daubois) [inline src/Symfony/Bridge/Doctrine/Tests/Decorator/DoctrineTransactionDecoratorTest.php]: You have places where you inline the `;`, and sometimes you don't, is there a reason?

Also, maybe these can be oneliners like this as they are pretty basic: `$this->entityManager->expects($this->once())->method('flush');`. But that's just personal taste so I'd understand if you don't like it!

File: src/Symfony/Bridge/Doctrine/Tests/Decorator/DoctrineTransactionDecoratorTest.php [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+        $this->connection->expects($this->once())
+            ->method('rollBack')
+        ;
+
+        $this->expectException(\RuntimeException::class);
+        $this->expectExceptionMessage('A runtime error.');
+
+        $handler = new TestHandler();
+        $this->decorator->call($handler->handleWithError(...));
```
Comment (alexandre-daubois) [inline src/Symfony/Bridge/Doctrine/Tests/Decorator/DoctrineTransactionDecoratorTest.php]: You may move the TestHandler instanciation above exception expectations. I know there are many places we do this, so you're sure that `new TestHandler()` is not the thing that throws and give a false-positive test.

(Same for below test)

File: src/Symfony/Bridge/Doctrine/Tests/Decorator/DoctrineTransactionDecoratorTest.php [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+    }
+
+    public function testInvalidEntityManagerThrowsException()
+    {
+        $this->managerRegistry
+            ->method('getManager')
+            ->with('unknown_manager')
+            ->will($this->throwException(new \InvalidArgumentException()));
```
Comment (alexandre-daubois) [inline src/Symfony/Bridge/Doctrine/Tests/Decorator/DoctrineTransactionDecoratorTest.php]: I think this notation is deprecated, you should use `willThrowException()` instead

File: src/Symfony/Bundle/FrameworkBundle/CHANGELOG.md [hunk 11]
```diff
@@ -12,6 +12,8 @@ CHANGELOG
  * Deprecate making `cache.app` adapter taggable, use the `cache.app.taggable` adapter instead
  * Enable `json_decode_detailed_errors` in the default serializer context in debug mode by default when `seld/jsonlint` is installed
  * Register `Symfony\Component\Serializer\NameConverter\SnakeCaseToCamelCaseNameConverter` as a service named `serializer.name_converter.snake_case_to_camel_case` if available
+ * Add Decorator component integration
+ * Add controller decoration support
```
Comment (alexandre-daubois) [inline src/Symfony/Bundle/FrameworkBundle/CHANGELOG.md]: Is this line necessary? It feels like in it already "bundled" in the line above?

**Review by alexandre-daubois — COMMENTED** [review]
New round :smile:

File: src/Symfony/Bundle/FrameworkBundle/CHANGELOG.md (hunk 11 above)
Comment (yceruto) [reply src/Symfony/Bundle/FrameworkBundle/CHANGELOG.md]: In this case the Framework integration involved enabling DI only (autowiring, autoconfiguration, compiler pass, etc.) allowing applications to effortlessly use the feature. The second entry focused on enabling the feature itself within the kernel/controller mechanism.

*We might want similar support for messenger/handlers in another PR*

File: src/Symfony/Bridge/Doctrine/Decorator/DoctrineTransactionDecorator.php (hunk 7 above)
Comment (yceruto) [reply src/Symfony/Bridge/Doctrine/Decorator/DoctrineTransactionDecorator.php]: This is the same implementation of `Doctrine\ORM\EntityManager::wrapInTransaction()`.

**Comment by yceruto** [comment]
I updated the PR description to include the Pros and Cons

**Comment by maxhelias** [comment]
Interesting! If I understand correctly, I see another useful decorators like “memoize” to cache the result of a method. Looking forward to seeing more feedback 😉

**Comment by yceruto** [comment]
The last commit simplifies decorator definition by allowing it to extend `Decorate` attribute and implement the `DecoratorInterface` at the same time (in an all-in-one approach) which is very convenient when your decorator doesn't require dependencies and doesn't need to be a service.

Example:
```php
#[\Attribute(\Attribute::TARGET_METHOD)]
class MyDecorator extends DecoratorAttribute implements DecoratorInterface
{
    public function decorate(\Closure $func): \Closure
    {
        return function (mixed ...$args) use ($func): mixed
        {
            echo "Do something before\n";

            $result = $func(...$args);

            echo "Do something after\n";

            return $result;
        };
    }

    public function decoratedBy(): string
    {
        return self::class;
    }
}

class GreetingController
{
    #[MyDecorator]
    public function sayHello(string $name): void
    {
        echo "Hello $name!\n";
    }
}
```

**Comment by yceruto** [comment]
Well, we could also apply this simple approach for decorator services by implementing `ServiceSubscriberInterface` + `ServiceMethodsSubscriberTrait` and leaving the constructor free for decorator options only:

```php
#[\Attribute(\Attribute::TARGET_METHOD)]
class MyDecorator extends DecoratorAttribute implements DecoratorInterface, ServiceSubscriberInterface
{
    use ServiceMethodsSubscriberTrait;

    public function __construct(public string $foo = 'bar')
    {
    }

    public function decorate(\Closure $func, self $metadata = new self()): \Closure
    {
        // do something with $this->someDependency()
 
        return function (mixed ...$args) use ($func, $metadata): mixed
        {
            echo "Do something with $metadata->foo before\n";

            $result = $func(...$args);

            echo "Do something with $metadata->foo after\n";

            return $result;
        };
    }

    public function decoratedBy(): string
    {
        return self::class;
    }

    #[SubscribedService]
    protected function someDependency(): SomeDependency
    {
        return $this->container->get(__METHOD__);
    }
}

class GreetingController
{
    #[MyDecorator(foo: 'baz')]
    public function sayHello(string $name): void
    {
        echo "Hello $name!\n";
    }
}
```

**Comment by yceruto** [comment]
Status: Improving proposal design

Still not happy with the way attribute options are passed through the decorator implementation, it feels a bit hacky and requires knowledge/documentation.
```php
parent::__construct(self::class, ['foo' => $foo]);
```

The way this function decoration works will require knowledge to wrap the function call:
```php
return function (mixed ...$args) use ($func): mixed { // <-- not obvious you've to do it like this to wrap the func call
    // return $func(...$args) 
}
```

I was looking into the Constraint/Validator design and I found it very similar to what I'd like to achieve here. So I'm going to make some changes to improve the DX.

**Comment by yceruto** [comment]
Proposal updated for better DX, please recheck PR description.

**Comment by ostrolucky** [comment]
I had a need for this couple times, but it was always with passing inheritance checks. This component is not doing that, it doesn't try to create a class that follows the hierarchy of decorated class, does it? I'm doing that eg. at https://github.com/snc/SncRedisBundle/blob/c095a0dadc6fd263e54e064f82b9dc4c8e7f9823/src/Factory/PhpredisClientFactory.php#L320 (thanks to proxy-manager). Without this, component is going to be much less useful, as you can't inject it in place whatever else was being injected before without changing the place where it's getting injected.

**Comment by yceruto** [comment]
@ostrolucky I understand that this might be a bit confusing at first, especially since we’re already familiar with a decoration method that addresses this kind of problem. However, this new component and its integration are intentionally designed to solve the decoration need in a different way for two main reasons:
1. **Support for multiple and nested decorators**: this approach allows for several decorators to be applied and even nested, providing greater flexibility in how we handle decorations.
2. **Reusability and decoupling**: by decoupling the decorator from the targeted method, we enhance reusability. Decorators no longer need to be aware of any specific class or method signature, making them more versatile.

In other words, instead of decorating the service using inheritance (as we used to do by making the decorator aware of the target class and methods), this decoration method targets any callable. This means that multiple decorators can be applied to any callable, and these decorators can be reused without needing to know the details of the classes or methods they’re decorating.

Of course, this approach is best suited for situations where this kind of decoration is beneficial, such as in handlers or controllers.

**Comment by RobinHoutevelts** [comment]
I (at least I think so) like this! 🤩 
However, I do think I might get a fatigue of repeating the same decorators.

```php
#[CollectCacheTags]
#[Serialize(format: 'json')]
#[Validate]
public function __invoke(#[MapRequestPayload] TaskPayload $payload): Task
```

Is there a way to combine these three to eg `#[JsonRequest]`, or should I go "back" to using regular decorators? The use-case above is a bad example, but hopefully my question makes sense. Or would a (future) improvement be to have "compound" decorators like we [have for constraints](https://symfony.com/doc/current/reference/constraints/Compound.html)?

**Comment by yceruto** [comment]
@RobinHoutevelts great idea! It’s definitely possible to add such a nice feature. In the meantime, I’m backporting it into [yceruto/decorator-bundle](https://github.com/yceruto/decorator-bundle) with support for `Compound` decorators already ;) thanks!

**Comment by norkunas** [comment]
I think this could be useful to integrate locking.

But if the key is dynamic would it be possible to use something similar like with `#[IsGranted(PostVoter::DELETE, subject: new Expression('args["postId"]'))]` where you can pass arguments to build the lock key?

**Comment by yceruto** [comment]
>I think this could be useful to integrate locking.
>
>But if the key is dynamic would it be possible to use something similar like with #[IsGranted(PostVoter::DELETE, subject: new Expression('args["postId"]'))] where you can pass arguments to build the lock key?

It's possible because you have control over the controller's arguments during the decoration phase. At this point, all arguments are already resolved, and their values are the same as what you'll receive in the controller method. However, it's harder to manipulate the arguments since they are just a simple list of values at this stage (not named arguments).

I think that for use cases where you want to add logic based on the controller’s arguments, using a `ControllerArgumentsEvent` listener is a better and simpler option.

I'd say that a Callable Decorator is a more generic & versatile solution for prepending or appending logic around controllers or any other callable, allowing you to combine decorators or reuse them for different callables. The `Transactional` decorator is one of the best examples, but I can see many other like caching responses, serializing results, rate limiting, error handling, etc.

**Comment by yceruto** [comment]
I added the Messenger/handler integration since the changes are minimal. Here's what this feature allows you to do:

```php
namespace App\Handler;

use Symfony\Bridge\Doctrine\Decorator\Transactional;

#[AsMessageHandler]
class ProductCreatedEventHandler
{
    #[Transactional]
    public function __invoke(ProductCreatedEvent $event): void
    {
        // update other related data...
    }
}
```

Similar to controllers, callable decorators are helpful when you need to decorate the implementation of specific handlers. However, if the decorator's logic applies to all handlers on the message bus, it's better to create a middleware instead.

**Comment by yceruto** [comment]
After an internal discussion about the component's name, we decided to rename it to "CallableWrapper". PR and description updated.

**Comment by yceruto** [comment]
I'm closing this for now. I'll revisit this topic soon via an RFC issue with more details on how to implement this feature in a way I believe is better than the one proposed here. Thanks all for your inputs!
